core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -820,24 +821,64 @@ void finishBecauseUnavailable(ShardId shardId, String message) {
         }
     }
 
+
+    static ConcurrentMap<IndexShardReference, String> openShardReferences;
```

Comment:
This feels way more natural on the IndexShard level - I felt that way  when I made it but couldn't find a way (at the time) to push it to IndexShard while getting the debugging information to be used as the string value here. We can pass it to IndexShard.acquirePrimaryOperationLock and acquireReplicaOperationLock but that will force a toString on the request object. How about making the reason an Object type and only call toString on it in IndexShard?